### PR TITLE
Rewrite exercise 0 to version2

### LINKED ISSUE #36: Rewrite exercise 0
Use typescript
Set up prettier 

please, in the version2 branch use the following link https://docs.google.com/presentation/d/1w8HRmAK2HB5PuwOB0HdFeGbvntsvM3-rvHdNOPZ3lBs/edit

please update the packages to the newest version (not simply npm update)

branching model:

version2 branches off master.
rewrite-exercise-0 branches off version2

### PR DESCRIPTION
**Pull request by dinassh**
The following is done in this PR: 
- rewrite exercise 0 using typescript
- remove default exports from exercise 0
- install and apply prettier

### CODE REVIEW

File: exercises/00-init/src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import 'babel-polyfill'
```
Comment (branoholy) [inline exercises/00-init/src/index.tsx]: Might not be necessary anymore. Anyway, it's deprecated https://babeljs.io/docs/en/babel-polyfill.

File: exercises/.prettierrc.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "singleQuote": true,
+  "semi": false,
```
Comment (branoholy) [inline exercises/.prettierrc.json:3]: I would go with semicolons. :slightly_smiling_face:

File: exercises/00-init/src/modules/root/components/root.tsx [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import React from 'react'
+
+export const Root = () => (
```
Comment (branoholy) [inline exercises/00-init/src/modules/root/components/root.tsx:3]: :+1: for removing default imports.

File: exercises/00-init/src/react-app-env.d.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+/// <reference types="react-scripts" />
```
Comment (branoholy) [inline exercises/00-init/src/react-app-env.d.ts:1]: Why is this necessary? I'm just curious because I don't use CRA very often. Actually, I haven't used it since this training. :smiley:

**Review by novellizator — COMMENTED** [review]
.

**Review by branoholy — CHANGES_REQUESTED** [review]
Good job! The only real problem is missing lock file.

File: exercises/.prettierrc.json (hunk 2 above)
Comment (novellizator) [reply exercises/.prettierrc.json:3]: please noooooo! :)

**Review by novellizator — COMMENTED** [review]
Just a comment for future - since we are using TS, we don't need to use proptypes in the following exercises

File: README.md [hunk 5]
```diff
@@ -1,8 +1,8 @@
 # Syllabus
-https://docs.google.com/presentation/d/1xaw0AXhimgnBBTctfTFHHIHddcI1RBUM0ITjexDiRRo
+https://docs.google.com/presentation/d/1w8HRmAK2HB5PuwOB0HdFeGbvntsvM3-rvHdNOPZ3lBs
 
 # Before you begin
-Install all dependencies and link executables to child projects with `yarn` or `npm i` in the `exercises` folder.
+Install all dependencies and link executables to child projects with `npm i` in the `exercises` folder.
```
Comment (novellizator) [inline README.md:5]: Since both me and @branoholy failed to run npm i on the first try, I'd suggest to move the package.json to the root directory. What do you think?

**Review by novellizator — COMMENTED** [review]
Well done! Please address the minor points ;)

File: exercises/00-init/src/react-app-env.d.ts (hunk 4 above)
Comment (dinassh) [reply exercises/00-init/src/react-app-env.d.ts:1]: It suppose to reference the react-script types to help with some imports.

Either way, it's automatically generated when I run `npm start`. Even if you delete the file it will be created again (possibly meaning, that it's important?).

I'm unsure how to proceed with this one.

File: exercises/00-init/src/modules/root/components/root.tsx (hunk 3 above)
Comment (dinassh) [reply exercises/00-init/src/modules/root/components/root.tsx:3]: :)

File: exercises/.prettierrc.json (hunk 2 above)
Comment (dinassh) [reply exercises/.prettierrc.json:3]: > please noooooo! :)

It's a strong opinion :smile: 

@branoholy Would you mind if the semicolons were removed?

File: exercises/00-init/src/react-app-env.d.ts (hunk 4 above)
Comment (branoholy) [reply exercises/00-init/src/react-app-env.d.ts:1]: OK, no problem to keep it. Or we could add it to `.gitignore` if it's generated because future versions of CRA might not need it.

File: exercises/.prettierrc.json (hunk 2 above)
Comment (branoholy) [reply exercises/.prettierrc.json:3]: I personally prefer semicolons so I would vote for them but I'm OK with both options. The most important thing is that it's set in the prettier config. :+1:

File: README.md (hunk 5 above)
Comment (dinassh) [reply README.md:5]: This will result in the same kind or problem as we encountered while updating `concurrently` package: the [link-parent-bin](https://github.com/nicojs/node-link-parent-bin) package will not update the binaries of parent dependencies as we would like. In this case it wont even try to add links in exercise directories (00-09) during `npm i`, as they are not located in neither `.` nor `packages` directory related to the root of the project. When overriding the child expected directory with `-c exercises` argument the binaries are not updated for those of us who cloned the old version (with package.json in exercises directory) and already installed the packages from there:
```
Linking dependencies ["concurrently","link-parent-bin","nodemon","prettier"] under children ["00-init","01-react-stateful","02-react-stateless","03-redux","04-react-redux","05-reselect","06-redux-saga","07-normalizr","08-router5","09-forms","node_modules"]
09:51:51 (7409) INFO link Different link at 'exercises/01-react-stateful/node_modules/.bin/prettier' already exists. Leaving it alone, the package is probably already installed in the child package.
...
```

Comment (karelklima) [reply README.md:5]: @novellizator @branoholy @dinassh  There is a lot of magic going on with the solution above. Considering that this is a training project, we should keep the non-react stuff down to the minimum.

I think that the idea of having a single package.json file is great, but linking dependencies with some magical symlinking not so much :-) For the sake of simplicity I suggest this:
1. Replace CRA with Parcel bundler
2. Have one package.json file in the root directory
3. Have multiple start commands specified, i.e. `start:0` would do `parcel src/00-init/index.html` etc.

With the above, we should have one set of dependencies but still be able to invoke the app from multiple entry points, which is what we want to do unless I am really mistaken.

AFAIK Parcel should do out of the box everything that CRA does, maybe except ESlint stuff.

What do you think?

File: exercises/00-init/src/index.tsx (hunk 1 above)
Comment (dinassh) [reply exercises/00-init/src/index.tsx]: I've removed the import from this file. See https://github.com/salsita/react-training/pull/37/commits/32c7621306683809d52044ca92e2d7e85c4b0c67

The package will be removed from package.json when all exercises are updated. I've created an issue for that, see https://github.com/salsita/react-training/issues/39

File: README.md (hunk 5 above)
Comment (dinassh) [reply README.md:5]: @karelklima FYI during today's discussion it was proposed to move all exercise directories (00-09) up to the root

1. I don't have much experience with either of them, so I don't have the preference at this point.
2. I propose to address this issue in another pull request.

Comment (novellizator) [reply README.md:5]: Good idea worth considering! We didn't plan to address it in this PR per se as it's low priority and can be done anytime. I created an issue for that here https://github.com/salsita/react-training/issues/40

**Review by novellizator — APPROVED** [review]
Happy with how it is!

**Review by branoholy — APPROVED** [review]
